### PR TITLE
Add http.Transport + performance improvements

### LINKED ISSUE #33: High amount of unclosed abandoned connections
### 1. What version of Caddy are you using (`caddy -version`)?
0.11.0

### 2. What are you trying to do?
use caddy as a forward proxy chain

### 3. What is your entire Caddyfile?
```text:24096 {
    forwardproxy {
        response_timeout 30
        dial_timeout     30
        ports     80 443
        upstream  https://user:mypass@xxx.mydomain.com:443
    }
    root /root/bin/caddy/html
    gzip
    errors /root/bin/caddy/1936/error.log
    log /root/bin/caddy/1936/access.log
}
```
### 4. How did you run Caddy (give the full command and describe the execution environment)?
ulimit -n 16384 && caddy
on debian 9
load balanced by haproxy ( with statics ), than export to internal user.

### 6. What did you expect to see?
a few ESTABLISHED connection to remote host after running a few hours

### 7. What did you see instead (give full error messages and/or log)?
a lot of connection 
netstat -an | grep 64.xxx.xx.xxx | grep ESTABLISHED| wc -l
123
on haproxy statics page, current session is only 17 and max session is 67.

### 8. How can someone who is starting from scratch reproduce the bug as minimally as possible?
sorry, no clue yet, I've try do curl every  15 secs,  but can not reproduce.

So I hope someone can explain upstream connection reuse and timeout rule for me.

### LINKED ISSUE #45: Add HTTP/2 support for connecting to upstream proxy server
Currently, if the upstream proxy server supports only HTTP/2, forwardproxy can not work with it.
Could you please add support for connecting to HTTP/2 upstream proxy server?
Thank you!

### PR DESCRIPTION
**Pull request by sergeyfrolov**
### 1. What does this change do, exactly?
Adds http.Transport to handle GET requests for non-upstreaming proxies (implements #37).
Upstreaming proxies will use httpclient to establish persistent connection to upstream instead.
Fix the issue with hang connections in `dualStream` (fixes #33)
httpclient: unlock the client before doing RoundTrip (fixes #45, more precisely, the issue that @joeshae reported in the last comment) 

### 2. Please link to the relevant issues.
implements #37
fixes #33
fixes #45

### 3. Which documentation changes (if any) need to be made because of this PR?
Seems to be none.

### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I made pull request as minimal and simple as possible. If change is not small or additional dependencies are required, I opened an issue to propose and discuss the design first
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code

### CODE REVIEW

File: forwardproxy.go [hunk 1]
```diff
@@ -99,25 +103,27 @@ func (fp *ForwardProxy) portIsAllowed(port string) bool {
 	return isAllowed
 }
 
-// Copies data r1->w1 and r2->w2, flushes as needed, and returns when both streams are done.
-func dualStream(w1 io.Writer, r1 io.Reader, w2 io.Writer, r2 io.Reader) error {
-	errChan := make(chan error)
+// Copies data targetWriter->clientReader and clientWriter->targetReader, and flushes as needed
+// Returns when both clientWriter-> targetReader stream is done. Caddy should finish writing
```
Comment (bemasc) [inline forwardproxy.go]: "both"?

File: forwardproxy.go [hunk 2]
```diff
@@ -99,25 +103,27 @@ func (fp *ForwardProxy) portIsAllowed(port string) bool {
...
+// targetWriter -> clientReader.
+func dualStream(targetWriter io.WriteCloser, clientReader io.ReadCloser,
```
Comment (bemasc) [inline forwardproxy.go]: `WriteCloser` seems wrong here, since we never call `Close`

File: forwardproxy.go [hunk 3]
```diff
@@ -395,21 +398,58 @@ func (fp *ForwardProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int,
...
-
+		var err error
 		var response *http.Response
-		err = r.Write(targetConn)
-		if err != nil {
-			return http.StatusBadGateway, errors.New("failed to write http request: " + err.Error())
+		if fp.upstream == nil {
+			// non-upstream request uses httpTransport to reuse connections
+			if r.Body != nil &&
+				(r.Method == "GET" || r.Method == "HEAD" || r.Method == "OPTIONS" || r.Method == " TRACE") {
```
Comment (bemasc) [inline forwardproxy.go]: Extra space in `" TRACE"`

File: forwardproxy.go [hunk 4]
```diff
@@ -395,21 +398,58 @@ func (fp *ForwardProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int,
...
+				// make sure request is idempotent and could be retried by saving the Body
```
Comment (bemasc) [inline forwardproxy.go]: None of these methods is allowed to have a body, so this comment is confusing.  I think it's fine to copy the body, but the comment should probably be clear that the body is normally expected to be empty.

File: forwardproxy.go [hunk 5]
```diff
@@ -395,21 +398,58 @@ func (fp *ForwardProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int,
...
+				r.GetBody = func() (io.ReadCloser, error) {
+					return ioutil.NopCloser(bytes.NewReader(rBodyBuf)), nil
+				}
+				r.Body, _ = r.GetBody()
+			}
+			response, err = fp.httpTransport.RoundTrip(r)
+		} else {
+			// Upstream requests don't interact well with Transport: connections could always be
+			// reused, but Transport thinks they go to different Hosts, so it spawns tons of
+			// useless connections. Just use dialContext, which will multiplex via single connection
```
Comment (bemasc) [inline forwardproxy.go]: Will `dialContext` multiplex if the upstream connection is HTTP/1.1?  The comment could be clearer.

File: httpclient/httpclient.go [hunk 6]
```diff
@@ -269,7 +274,14 @@ func (h *http2Conn) Write(p []byte) (n int, err error) {
...
 
 func (h *http2Conn) Close() error {
-	h.out.Close()
 	h.in.Close()
-	return h.Conn.Close()
+	return h.out.Close()
+}
+
+func (h *http2Conn) CloseWrite() error {
+	return h.out.Close()
```
Comment (bemasc) [inline httpclient/httpclient.go]: This looks backwards, given that `h.out` is a `Reader`, not a `Writer`...

File: forwardproxy.go (hunk 1 above)
Comment (sergeyfrolov) [reply forwardproxy.go]: Fixed.

File: forwardproxy.go (hunk 2 above)
Comment (sergeyfrolov) [reply forwardproxy.go]: `ReadCloser`s are also wrong, we don't call close on them either.
dualStream function is no longer asymmetric, and this argument could help to catch errors if arguments are swapped.

File: forwardproxy.go (hunk 3 above)
Comment (sergeyfrolov) [reply forwardproxy.go]: Fixed

File: forwardproxy.go (hunk 4 above)
Comment (sergeyfrolov) [reply forwardproxy.go]: Added comment

File: forwardproxy.go (hunk 5 above)
Comment (sergeyfrolov) [reply forwardproxy.go]: Fixed

File: httpclient/httpclient.go (hunk 6 above)
Comment (sergeyfrolov) [reply httpclient/httpclient.go]: Fixed

File: forwardproxy.go [hunk 7]
```diff
@@ -99,25 +103,27 @@ func (fp *ForwardProxy) portIsAllowed(port string) bool {
 	return isAllowed
 }
 
-// Copies data r1->w1 and r2->w2, flushes as needed, and returns when both streams are done.
-func dualStream(w1 io.Writer, r1 io.Reader, w2 io.Writer, r2 io.Reader) error {
-	errChan := make(chan error)
+// Copies data targetWriter->clientReader and clientWriter->targetReader, and flushes as needed
+// Returns when clientWriter-> targetReader stream is done. Caddy should finish writing
+// targetWriter -> clientReader.
+func dualStream(targetWriter io.WriteCloser, clientReader io.ReadCloser,
```
Comment (bemasc) [inline forwardproxy.go]: Per our conversation, maybe just make this `target net.Conn`.

File: forwardproxy.go [hunk 8]
```diff
@@ -103,13 +103,13 @@ func (fp *ForwardProxy) portIsAllowed(port string) bool {
...
 
-// Copies data targetWriter->clientReader and clientWriter->targetReader, and flushes as needed
-// Returns when clientWriter-> targetReader stream is done. Caddy should finish writing
-// targetWriter -> clientReader.
-func dualStream(targetWriter io.WriteCloser, clientReader io.ReadCloser,
-	clientWriter io.Writer, targetReader io.ReadCloser) error {
+// Copies data target->clientReader and clientWriter->target, and flushes as needed
+// Returns when clientWriter-> target stream is done.
+// Caddy should finish writing target -> clientReader.
+func dualStream(target net.Conn, clientReader io.ReadCloser,
```
Comment (bemasc) [inline forwardproxy.go]: Nit: I think `Reader` might make more sense than `ReadCloser`.

**Pull request merged**